text/0000-import-trait-associated-functions.md

```diff
@@ -0,0 +1,235 @@
+        S {
+            a: Default::default()
+        }
+    }
+}
+```
+And a call
+```rust
+use Trait::func as m;
+m(x, y, z);
+```
+desugars to
+```rust
+Trait::func(x, y, z);
+```
+
+Additionally, the syntax
+```rust
+use some_module::Trait::self;
+```
+is sugar for
+```rust
+use some_module::Trait;
+```
+to allow importing a trait and its methods at the same time, e.g:
+```rust
+use Default::{self, default};
+```
+
+The restriction on importing parent trait associated functions is a consequence of this desugaring, see https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=51bef9ba69ce1fc20248e987bf106bd4 for examples of the errors you get when you try to call parent trait associated functions through a child trait. We will likely want better error messages than this if a user tries to import a parent function.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Calls to `default` are less explicit than calls to `Default::default` or to `T::default`, likewise for any other trait. Some users may see this lack of explicitness as bad style.
+
+To expand on this, [the book](https://doc.rust-lang.org/book/ch07-04-bringing-paths-into-scope-with-the-use-keyword.html#creating-idiomatic-use-paths) currently recommends that functions should be called using their parent module's name:
+> Although both Listing 7-11 and 7-13 accomplish the same task, Listing 7-11 is the idiomatic way to bring a function into scope with use. Bringing the function’s parent module into scope with use means we have to specify the parent module when calling the function. Specifying the parent module when calling the function makes it clear that the function isn’t locally defined while still minimizing repetition of the full path.
+
+This recommendation makes the most sense when there is a possibility of ambiguity in the mind of the reader. For example, a function like `sin` is unlikely to be ambiguous, because there is only one mathematical function of that name. If a codebase is likely to be making use of multiple different implementations of `sin`, then it makes more sense to require specifically naming the one you are going to use. Similar considerations apply to traits like `Default::default`, or more generally in cases like `Frobnicator::frobnicate`.
+
+Because of this context sensitivity, we should allow developers to choose when removing the extra context makes sense for their codebase.
+
+Another drawback mentioned during review for this RFC was that this adds more complication to the name resolution rules. On an implementation side, I am assured that this feature is straightforward to implement. From a user perspective, the name lookup rules for the function name are exactly the same as those used to look up any other function name. The lookup rules used to resolve the `impl` are also exactly the same ones used for non-fully qualified trait function calls. There is no fundamentally new kind of lookup happening here, just a remixing of existing lookup rules.
```

Comment:
> On an implementation side, I am assured that this feature is straightforward to implement.

Note that this may be the case for rustc, but not necessarily for alternative implementations. Wearing my rust-analyzer hat I'd expect this to be doable as well, but a bit annoying. Not saying this is a blocker (it's not impossible to implement) but mainly to keep that in mind until we attempt to implement it in r-a.

Reply:
I considered writing a section with some considerations on UX for IDEs. This does bring in a new level of complexity for auto-imports in IDEs. I decided to keep this focused on rustc and the core language however.